File: 2023/d17.R
parse_input <- function(path) {
  s <- readLines(path)
  matrix(as.integer(unlist(strsplit(s, NULL))), length(s), byrow = TRUE)
}


translate <- function(n) {
  n2 <- n * n
  list(encode = function(r, c, d) r + (c - 1L) * n + (d - 1L) * n2,
       row = function(x) (x - 1L) %% n + 1L,
       col = function(x) (x - 1L) %% n2 %/% n + 1L,
       dir = function(x) (x - 1L) %/% n2 + 1L)
}


best_path <- function(grid, min_steps, max_steps) {
  ##             west       south      east        north
  move <- list(c(0L, 1L), c(1L, 0L), c(0L, -1L), c(-1L, 0L))
  turn <- list(c(4L, 2L), c(1L, 3L), c(2L, 4L),  c(3L, 1L))
  n <- nrow(grid)
  tr <- translate(n)
  dist <- rep(Inf, n * n * 4)
  ## A very poor priority queue!
  value <- c(0, 0)
  data <- c(tr$encode(1, 1, 1), tr$encode(1, 1, 2))
  
  repeat {
    idx <- which.min(value)
    cost <- value[[idx]]
    x <- data[[idx]]
    value[idx] <- Inf
    r <- tr$row(x)
    c <- tr$col(x)
    d <- tr$dir(x)
    if (r == n && c == n) {
      return(cost)
    }
    if (cost > dist[[x]]) {
      next
    }
    for (d_next in turn[[d]]) {
      delta <- move[[d_next]]
      cost_next <- cost
      for (i in seq_len(max_steps)) {
        r_next <- r + delta[[1]] * i
        c_next <- c + delta[[2]] * i
        if (r_next < 1 || r_next > n || c_next < 1 || c_next > n) {
          break
        }
        cost_next <- cost_next + grid[r_next, c_next]
        if (i <= min_steps) { # 1, 2, 3 rejected for min_steps = 3
          next
        }
        x_next <- tr$encode(r_next, c_next, d_next)
        if (cost_next < dist[[x_next]]) {
          dist[[x_next]] <- cost_next
          value[[length(value) + 1]] <- cost_next
          data[[length(data) + 1]] <- x_next
        }
      }
    }
  }
}


part1 <- function(grid) {
  best_path(grid, 0, 3)
}


part2 <- function(grid) {
  best_path(grid, 3, 10)
}


options(error=recover)
test <- parse_input("test-input-d17.txt")
stopifnot(part1(test) == 102)
stopifnot(part2(test) == 94)

d <- parse_input("input-d17.txt")
part1(d) # 1260
part2(d) # 1416

File: 2023/d17.py
from collections import defaultdict
from heapq import heappop, heappush
from math import inf

def parse_input(path):
    with open(path) as f:
        return [list(map(int, line.strip())) for line in f.readlines()]


def best_path(grid, min_steps, max_steps):
    move = {"w": (0, 1), "s": (1, 0), "e": (0, -1), "n": (-1, 0)}
    turn = {"w": ("n", "s"), "s": ("w", "e"), "e": ("s", "n"), "n": ("e", "w")}
    n = len(grid)
    dist = defaultdict(lambda: inf)
    heap = [(0, (0, 0, "w")), (0, (0, 0, "s"))]
    while heap:
        cost, (r, c, d) = heappop(heap)
        if r == n - 1 and c == n - 1:
            return cost
        if cost > dist[r, c, d]:
            continue
        for d_next in turn[d]:
            dr, dc = move[d_next]
            cost_next = cost
            for i in range(max_steps):
                r_next, c_next = r + dr * (i + 1), c + dc * (i + 1)
                if r_next < 0 or r_next >= n or c_next < 0 or c_next >= n:
                    break
                cost_next += grid[r_next][c_next]
                if i < min_steps: # 0, 1, 2 get rejected for min_steps = 3
                    continue
                p_next = (r_next, c_next, d_next)
                if cost_next < dist[p_next]:
                    dist[p_next] = cost_next
                    heappush(heap, (cost_next, p_next))


def part1(grid):
    return best_path(grid, 0, 3)


def part2(grid):
    return best_path(grid, 3, 10)


test = parse_input("test-input-d17.txt")
assert part1(test) == 102
assert part2(test) == 94

d = parse_input("input-d17.txt")
print(part1(d)) # 1260
print(part2(d)) # 1416
